src/cargo/ops/cargo_package.rs

```diff
@@ -1157,7 +1157,7 @@ fn run_verify(
                 opts.jobs.clone(),
                 opts.keep_going,
                 &opts.targets,
-                CompileMode::Build,
+                CompileMode::Check { test: false },
```

Comment:
For myself I would keep it a full build.
One reason is that `cargo publish` is not an operation easy to revert.

The other is, per [RFC 3477](https://github.com/rust-lang/rfcs/blob/master/text/3477-cargo-check-lang-policy.md),

> A Rust program must compile with cargo build to be covered by Rust's standard stability guarantee.

If we want the same level of stability guarantee for published crates, we'd better do a full build.

Reply:
I guess the question is what is it we intend to be verified and at what cost.

Things that can be verified
- Basic check
- Post-monomorphization errors
- feature combinations
- platform combinations
- logic (ie `cargo test`, #14685)
- Miri for verifying `unsafe`

In commonly used crates, there is most likely a CI checking for what the author feels is important enough.  I don't think `cargo publish`, especially be default, should or can replicate that.

What verify can help with is helping specifically with packaging specific issues
- Are all the right files present
- Can this build in isolation (to a degree)

For those, `check` is sufficient.

For people who don't have a CI, verify can tell them some but not much.  Personally, I don't think we need to be verifying the stability guarantees in these cases, especially at the cost of expensive packaging, which has the most impact when doing it in dry-run mode which is where I most notice this (and tend to avoid dry run because of it).

Reply:
On one hand, commonly used crates usually tend to have better CI checking support, so less likely to hit errors like this. Yet it depends on where they run `cargo build` or `cargo test` under release profile.

On the other hand, we are getting more API in const contexts (81 in 1.83). While `const fn` doesn't always contribute to error during [monomorphization](https://github.com/rust-lang/rust/issues/99682), it still increases the chances of hitting it.
(Check the minimal example in <https://github.com/rust-lang/rust/issues/112301> to understand how to get bitten).

There are also [diagnostics during MIR passes](https://github.com/rust-lang/rust/issues/49292) cannot be caught by `cargo check`. By just looking at how easy [the example](https://github.com/rust-lang/rust/issues/49292#issuecomment-1691455732) is. I am afraid of this may be more troublesome than monomorphization errors.

> I searched through the issues, open and closed, and saw no previous discussions of this.

Supposedly this is a starting point of examining all verification Cargo provides and re-position cargo commands, to embrace the potential plumbing commands reorganization (https://github.com/rust-lang/rust-project-goals/pull/178/ perhaps). However, instead of merging into this change now, perhaps we could create an issue and solicit feedback first?

Reply:
By all means, I love how you’re thinking ahead with all these verification categories.

Reply:
> Yet it depends on where they run cargo build or cargo test under release profile.

Good point that there is also `--release`

> While const fn doesn't always contribute to error during https://github.com/rust-lang/rust/issues/99682, it still increases the chances of hitting it.

> There are also https://github.com/rust-lang/rust/issues/49292 cannot be caught by cargo check

If post-`check` errors are getting that bad, then I have strong words for T-lang...  The RFC test focuses on linker errors and doesn't justify its existence with it being prevalent.

Reply:
Created #14941